google/cloud/internal/big_endian.h

```diff
@@ -15,164 +15,69 @@
-#else
-    std::uint64_t return_value = ((((num_value)&0xff00000000000000ull) >> 56) |
-                                  (((num_value)&0x00ff000000000000ull) >> 40) |
-                                  (((num_value)&0x0000ff0000000000ull) >> 24) |
-                                  (((num_value)&0x000000ff00000000ull) >> 8) |
-                                  (((num_value)&0x00000000ff000000ull) << 8) |
-                                  (((num_value)&0x0000000000ff0000ull) << 24) |
-                                  (((num_value)&0x000000000000ff00ull) << 40) |
-                                  (((num_value)&0x00000000000000ffull) << 56));
-
-    return return_value;
-#endif
+// Encodes an 8-bit, 16-bit, 32-bit, or 64-bit signed or unsigned value as a
+// big-endian sequence of bytes. The returned string has a size matching
+// `sizeof(T)`. Example:
+//
+//   std::string s = EncodeBigEndian(std::int32_t{255});
+//   assert(s == std::string("\0\0\0\xFF", 4));
+//
+template <typename T,
+          typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+std::string EncodeBigEndian(T value) {
+  static_assert(std::numeric_limits<unsigned char>::digits == 8,
+                "This code assumes an 8-bit char");
+  using unsigned_type = typename std::make_unsigned<T>::type;
+  unsigned_type const n = value;
+  auto shift = sizeof(n) * 8;
+  std::array<unsigned char, sizeof(n)> a;
+  for (auto& c : a) {
+    shift -= 8;
+    c = (n >> shift) & 0xFF;
   }
+  return {a.begin(), a.end()};
+}
 
-  template <typename IntegralType>
-  static inline IntegralType as_big_endian(IntegralType native) {
-    return byte_swap(native);
+// Decodes the given string as a big-endian sequence of bytes representing an
+// integer of the specified type. The allowed types are std::int8_t through
+// std::int64_t, both signed and unsigned. Returns an error status if the given
+// string is the wrong size for the specified type. Example:
+//
+//   std::string s("\0\0\0\xFF", 4);
+//   StatusOr<std::int32_t> decoded = DecodeBigEndian(s);
+//   if (decoded) assert(*decoded == 255);
+//
+template <typename T,
+          typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+StatusOr<T> DecodeBigEndian(std::string const& value) {
+  static_assert(std::numeric_limits<unsigned char>::digits == 8,
+                "This code assumes an 8-bit char");
+  if (value.size() != sizeof(T)) {
+    auto const msg = "Given value with " + std::to_string(value.size()) +
+                     " bytes; expected " + std::to_string(sizeof(T));
+    return Status(StatusCode::kInvalidArgument, msg);
   }
-
-  template <typename IntegralType>
-  static inline IntegralType as_native(IntegralType big_endian) {
-    return byte_swap(big_endian);
+  using unsigned_type = typename std::make_unsigned<T>::type;
+  auto shift = sizeof(T) * 8;
+  T result = 0;
+  for (auto const& c : value) {
+    shift -= 8;
+    result |= (c & unsigned_type{0xFF}) << shift;
```

Comment:
It seems like this might also have implementation-defined behavior on the first loop when T is signed and c has the high bit on.